provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/DefaultProjectSchemaProvider.kt

```diff
@@ -0,0 +1,120 @@
+import org.gradle.api.reflect.HasPublicType
+import org.gradle.api.reflect.TypeOf
+import org.gradle.api.tasks.SourceSet
+
+import org.gradle.api.internal.HasConvention
+
+import org.gradle.kotlin.dsl.provider.spi.ProjectSchema
+import org.gradle.kotlin.dsl.provider.spi.ProjectSchemaEntry
+import org.gradle.kotlin.dsl.provider.spi.ProjectSchemaProvider
+
+
+class DefaultProjectSchemaProvider : ProjectSchemaProvider {
+
+    override fun schemaFor(project: Project): ProjectSchema<TypeOf<*>> =
+        targetSchemaFor(project, TypeOf.typeOf(Project::class.java)).let { targetSchema ->
+            ProjectSchema(
+                targetSchema.extensions,
+                targetSchema.conventions,
+                accessibleConfigurations(project.configurations.names.toList()))
+        }
+}
+
+
+private
+data class ExtensionConventionSchema(
+    val extensions: List<ProjectSchemaEntry<TypeOf<*>>>,
+    val conventions: List<ProjectSchemaEntry<TypeOf<*>>>
+)
+
+
+private
+fun targetSchemaFor(target: Any, targetType: TypeOf<*>): ExtensionConventionSchema {
+
+    val extensions = mutableListOf<ProjectSchemaEntry<TypeOf<*>>>()
+    val conventions = mutableListOf<ProjectSchemaEntry<TypeOf<*>>>()
+
+    fun collectSchemaOf(target: Any, targetType: TypeOf<*>) {
+        if (target is ExtensionAware) {
+            accessibleExtensionsSchema(target.extensions.extensionsSchema).forEach { schema ->
+                extensions.add(ProjectSchemaEntry(targetType, schema.name, schema.publicType))
+                if (!schema.isDeferredConfigurable) {
+                    collectSchemaOf(target.extensions.getByName(schema.name), schema.publicType)
+                }
+            }
+        }
+        if (target is HasConvention) {
+            accessibleConventionsSchema(target.convention.plugins).forEach { name, type ->
+                conventions.add(ProjectSchemaEntry(targetType, name, type))
+                collectSchemaOf(target.convention.plugins[name]!!, type)
+            }
+        }
+        if (target is Project) {
+            target.convention.findPlugin(JavaPluginConvention::class.java)?.sourceSets?.forEach { sourceSet ->
+                collectSchemaOf(sourceSet, sourceSetType)
+            }
+        }
+    }
+
+    collectSchemaOf(target, targetType)
+
+    return ExtensionConventionSchema(extensions.distinct(), conventions.distinct())
+}
+
+
+private
+fun accessibleExtensionsSchema(extensionsSchema: ExtensionsSchema) =
+    extensionsSchema.filter { isPublic(it.name) }
+
+
+private
+fun accessibleConventionsSchema(plugins: Map<String, Any>) =
+    plugins.filterKeys(::isPublic).mapValues { inferPublicTypeOfConvention(it.value) }
+
+
+private
+fun inferPublicTypeOfConvention(instance: Any) =
+    if (instance is HasPublicType) instance.publicType
+    else TypeOf.typeOf(instance::class.java.undecorated)
+
+
+private
+val Class<*>.undecorated
+    get() = takeIf { it.name.endsWith("_Decorated") }?.superclass ?: this
```

Comment:
:x: `name.endsWith("_Decorated")` seems very fragile. Maybe make the Gradle decorator add some marker interface or annotation to the types so they can be properly detected?